clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java

```diff
@@ -17,160 +17,78 @@
-    // have existed on the broker or not.
-    private static final Schema LEADER_AND_ISR_REQUEST_V1 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            PARTITION_STATES_V1,
-            LIVE_LEADERS_V0);
-
-    // LEADER_AND_ISR_REQUEST_V2 added a broker_epoch Field. This field specifies the generation of the broker across
-    // bounces. It also normalizes partitions under each topic.
-    private static final Schema LEADER_AND_ISR_REQUEST_V2 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            BROKER_EPOCH,
-            TOPIC_STATES_V2,
-            LIVE_LEADERS_V0);
-
-    // LEADER_AND_ISR_REQUEST_V3 added two new fields - adding_replicas and removing_replicas.
-    // These fields respectively specify the replica IDs we want to add or remove as part of a reassignment
-    private static final Schema LEADER_AND_ISR_REQUEST_V3 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            BROKER_EPOCH,
-            TOPIC_STATES_V3,
-            LIVE_LEADERS_V0);
-
-    public static Schema[] schemaVersions() {
-        return new Schema[]{LEADER_AND_ISR_REQUEST_V0, LEADER_AND_ISR_REQUEST_V1, LEADER_AND_ISR_REQUEST_V2, LEADER_AND_ISR_REQUEST_V3};
-    }
 
     public static class Builder extends AbstractControlRequest.Builder<LeaderAndIsrRequest> {
-        private final Map<TopicPartition, PartitionState> partitionStates;
+
+        private final List<LeaderAndIsrPartitionState> partitionStates;
         private final Set<Node> liveLeaders;
 
         public Builder(short version, int controllerId, int controllerEpoch, long brokerEpoch,
-                       Map<TopicPartition, PartitionState> partitionStates, Set<Node> liveLeaders) {
+                       List<LeaderAndIsrPartitionState> partitionStates, Set<Node> liveLeaders) {
             super(ApiKeys.LEADER_AND_ISR, version, controllerId, controllerEpoch, brokerEpoch);
             this.partitionStates = partitionStates;
             this.liveLeaders = liveLeaders;
         }
 
         @Override
         public LeaderAndIsrRequest build(short version) {
-            return new LeaderAndIsrRequest(controllerId, controllerEpoch, brokerEpoch, partitionStates, liveLeaders, version);
+            List<LeaderAndIsrLiveLeader> leaders = liveLeaders.stream().map(n -> new LeaderAndIsrLiveLeader()
+                .setBrokerId(n.id())
+                .setHostName(n.host())
+                .setPort(n.port())
+            ).collect(Collectors.toList());
+
+            LeaderAndIsrRequestData data = new LeaderAndIsrRequestData()
+                .setControllerId(controllerId)
+                .setControllerEpoch(controllerEpoch)
+                .setBrokerEpoch(brokerEpoch)
+                .setLiveLeaders(leaders);
+
+            if (version >= 2) {
+                Map<String, List<LeaderAndIsrPartitionState>> groupedMap = groupByTopic(partitionStates);
```

Comment:
hmm.  groupByTopic already has to iterate over every partitionState, so it can just build a `Map<String, LeaderAndIsrTopicState>` and then do `data.setTopicStates(map.values())`, right?

Reply:
That's reasonable. I was previously relying on a more general group by method, but now that I've moved it to this class, it makes sense.

Reply:
Note that we have to do `new ArrayList<>(map.values())`.